warehouse/query-core/src/main/java/datawave/query/jexl/visitors/AddShardsAndDaysVisitor.java

```diff
@@ -0,0 +1,120 @@
+package datawave.query.jexl.visitors;
+
+import com.google.common.base.Joiner;
+import com.google.common.collect.Lists;
+import datawave.query.Constants;
+import datawave.query.jexl.JexlASTHelper;
+import datawave.query.jexl.JexlNodeFactory;
+import datawave.util.StringUtils;
+import org.apache.commons.jexl2.parser.ASTAssignment;
+import org.apache.commons.jexl2.parser.ASTJexlScript;
+import org.apache.commons.jexl2.parser.ASTReference;
+import org.apache.commons.jexl2.parser.JexlNode;
+import org.apache.commons.jexl2.parser.JexlNodes;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.stream.Collectors;
+
+/**
+ * When SHARDS_AND_DAYS hints are supplied via the OPTIONS function, they should be added back into the query tree either by updating an existing
+ * SHARDS_AND_DAYS assignment node if one exists, or by creating and appending a new SHARDS_AND_DAYS assignment node into the query tree structure.
+ */
+public class AddShardsAndDaysVisitor extends RebuildingVisitor {
+    
+    private static final Joiner JOINER = Joiner.on(',');
+    
+    @SuppressWarnings("unchecked")
+    public static <T extends JexlNode> T update(JexlNode node, String shardsAndDays) {
+        if (node == null) {
+            return null;
+        }
+        
+        T copy = (T) copy(node);
```

Comment:
Do we really need to make a copy?  We can return the original tree if no changes are made, and the node.jjtAccept(visitor, node) call below will make a copy.  Am I missing something here?

Reply:
Good point, I have removed the unnecessary copy.